tests/std/tests/GH_002299_implicit_sfinae_constraints/test.compile.pass.cpp

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <memory>
+#include <type_traits>
+#include <version>
+
+#ifdef __cpp_lib_optional

```

Comment:
While this is conforming, our tests conventionally use the coarse-grained (and implementation-specific, of course) `_HAS_CXX17` etc. macros, when the presence of a feature doesn't vary in other ways (e.g. varying with compilers). The product code also follows this convention (e.g. we had a lot of fine-grained tests for `__cpp_lib_constexpr_dynamic_alloc` that were eventually unified into `_HAS_CXX20`). It's not a big deal, but preferring the coarse-grained macros makes it a bit easier to see what will be tested as the Standard mode varies, and it makes the remaining uses of the fine-grained feature-test macros stand out as being special.